### ISSUE TITLE
Allow Teleport to work with FIPS endpoints in AWS

### ISSUE THREAD

**Issue by pschisa** [gravitational/teleport#11291]
### What
AWS S3 FIPS endpoint(https://aws.amazon.com/compliance/fips/) only responds to virtual hosted-style addressing (https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html) according to their documentation.

It does not seem like Teleport allows for virtual hosting (https) format to reference S3 buckets.  When using the traditional S3 format we fail with the attached errors to upload.

Request is to allow Teleport to accept virtual hosted-style addressing to enable Teleport to use FIPS compatible endpoints.

I have not been able to identify a workaround yet.

~~~
Mar 21 16:09:10 Test-Cluster-Node1-US-East teleport[1432]: ERROR REPORT:
Mar 21 16:09:10 Test-Cluster-Node1-US-East teleport[1432]: Original Error: *trace.ConnectionProblemError timeout waiting for stream status update
Mar 21 16:09:10 Test-Cluster-Node1-US-East teleport[1432]: Stack Trace:
Mar 21 16:09:10 Test-Cluster-Node1-US-East teleport[1432]: /go/src/github.com/gravitational/teleport/lib/events/filesessions/fileasync.go:501 github.com/gravitational/teleport/lib/events/filesessions.(*Uploader).upload
Mar 21 16:09:10 Test-Cluster-Node1-US-East teleport[1432]: /go/src/github.com/gravitational/teleport/lib/events/filesessions/fileasync.go:434 github.com/gravitational/teleport/lib/events/filesessions.(*Uploader).startUpload.func1
Mar 21 16:09:10 Test-Cluster-Node1-US-East teleport[1432]: /opt/go/src/runtime/asm_amd64.s:1581 runtime.goexit
Mar 21 16:09:10 Test-Cluster-Node1-US-East teleport[1432]: User Message: timeout waiting for stream status update] filesessions/fileasync.go:435
Mar 21 16:09:10 Test-Cluster-Node1-US-East teleport[1432]: 2022-03-21T16:09:10Z WARN [UPLOAD]    Backing off, will retry after 10s. error:[
Mar 21 16:09:10 Test-Cluster-Node1-US-East teleport[1432]: ERROR REPORT:
Mar 21 16:09:10 Test-Cluster-Node1-US-East teleport[1432]: Original Error: *trace.ConnectionProblemError timeout waiting for stream status update
Mar 21 16:09:10 Test-Cluster-Node1-US-East teleport[1432]: Stack Trace:
Mar 21 16:09:10 Test-Cluster-Node1-US-East teleport[1432]: /go/src/github.com/gravitational/teleport/lib/events/filesessions/fileasync.go:501 github.com/gravitational/teleport/lib/events/filesessions.(*Uploader).upload
Mar 21 16:09:10 Test-Cluster-Node1-US-East teleport[1432]: /go/src/github.com/gravitational/teleport/lib/events/filesessions/fileasync.go:434 github.com/gravitational/teleport/lib/events/filesessions.(*Uploader).startUpload.func1
Mar 21 16:09:10 Test-Cluster-Node1-US-East teleport[1432]: /opt/go/src/runtime/asm_amd64.s:1581 runtime.goexit
Mar 21 16:09:10 Test-Cluster-Node1-US-East teleport[1432]: User Message: timeout waiting for stream status update] filesessions/fileasync.go:209
~~~

**Comment by pschisa** [comment]
Identified workaround to force Teleport to use FIPS without requiring explicitly setting FIPS in Teleport via virtual hosted-style addressing

~~~
we do use the fips endpoint if the host's ~/.aws/config specifies that it should be used, which seems like it should be the best-practice to do.  Secondly, it appears that the AWS SDK actually recognizes this env argument:
AWS_USE_FIPS_ENDPOINT=true
So for those that don't have fips endpoints enabled for the entire host for some reason can use that.

basing this off of the docs [here](https://docs.aws.amazon.com/sdk-for-go/api/aws/session/).  Specifically the "FIPS and DualStack Endpoints" section.
~~~

**Comment by r0mant** [comment]
@pschisa Does the customer run Teleport with `--fips` flag? AWS config has a `UseFIPSEndpoint` property. So instead of adding support for https URLs we could make Teleport set this flag when it's running in FIPS mode. I'm guessing it's the analogue of the environment variable you found. Would that work?

cc @alistanis

**Comment by alistanis** [comment]
@r0mant If we set `UseFIPSEndpoint` will we break existing customers that use `--fips` that, for whatever reason, do not expect to use FIPS in AWS/S3?

**Comment by r0mant** [comment]
@alistanis Yes, I've also been thinking about this. I don't know if we'll break anything but it might be an unexpected change. It won't also allow to use only, say, S3 FIPS endpoint without running teleport in FIPS mode.

So probably enabling it on the endpoint level would be a better approach for now - we can use the same query string style parameters we already use for other things (https://goteleport.com/docs/setup/reference/backends/#s3).

```
audit_sessions_uri: "s3://Example_TELEPORT_S3_BUCKET/records?use_fips_endpoint=true"
```

Take a look at https://github.com/gravitational/teleport/issues/13222 in the context of this also, we might need to do the same for DynamoDB.

**Comment by r0mant** [comment]
@alistanis That said, using FIPS endpoints when Teleport is running in FIPS mode is what a user would probably expect to be the default behavior. Otherwise they would need to know to toggle additional FIPS settings on S3/Dynamo endpoints. The only concern is backwards compatibility.

Since we're nearing Teleport 10 release, how about this:

- We add ability to enable/disable FIPS endpoints for S3 and Dynamo endpoints individually like discussed above.
- For master/v10, we make sure that Teleport enables FIPS endpoints automatically when running with `--fips`.
  - We'll document this as a potential breaking change in release notes and provide a way to disable it using options above.
- When backporting to v9, backport only S3/Dynamo settings and keep `--fips` flag behavior the same.

What do you think?

Additionally, the customer had concerns that even with `AWS_USE_FIPS_ENDPOINT` workaround they saw DNS requests for non-FIPS endpoints. Let's confirm that FIPS endpoints are indeed being used with that flag.

Also, even before all that - let's see if we can use the existing `?endpoint=xxx` query string parameter style variable to force the FIPS endpoint - if that works, we can offer that as workaround.

**Comment by alistanis** [comment]
@r0mant Yeah. I agree it makes sense to be the default and that the only concern is backwards compatibility 100%.

I'm testing this with the query parameters and wireshark with a dns filter. Unless the customer is using [DNSSEC](https://en.wikipedia.org/wiki/Domain_Name_System_Security_Extensions) and their own DNS that should be sufficient to capture any odd behavior there. I'll report back on that once I have some data to show.

As for the other points:

* Agree, enable individual endpoints for both.
* I think this makes sense. I am not a compliance expert but I think FIPS is an all in certification, and it seems easier to guarantee behavior with fewer options.
* Backporting only the optional stuff makes perfect sense.

**Comment by pschisa** [comment]
When testing originally with the format `audit_sessions_uri: "s3://example/recordings/?endpoint=s3-fips.us-east-1.amazonaws.com"`, the session recordings were only saved to the S3 bucket when the `AWS_USE_FIPS_ENDPOINT=true` env variable was also set.

This also led to the investigation of the AWS S3 FIPS docs which implied virtual style hosting is the only way to reference these endpoints

~~~
Amazon Simple Storage Service (S3)

Note: These Endpoints can only be used with Virtual Hosted-Style addressing. For example: https://bucket.s3-fips.us-east-2.amazonaws.com.
~~~

**Comment by alistanis** [comment]
Unfortunately using the query parameters does not work alone, as @pschisa stated, as it interpolates the URI `s3://ccooper-fips-bucket/audit_sessions?endpoint=s3-fips.us-east-1.amazonaws.com` as `s3-fips.us-east-1.amazonaws.com/bucket`. which is incorrect:

```
2022-06-10T00:49:49Z ERRO [S3]        "Failed to ensure that bucket \"ccooper-fips-bucket\" exists (RequestError: send request failed\ncaused by: Head \"https://s3-fips.us-east-1.amazonaws.com/ccooper-fips-bucket\": dial tcp: lookup s3-fips.us-east-1.amazonaws.com: No address associated with hostname). S3 session uploads may fail. If you've set up the bucket already and gave Teleport write-only access, feel free to ignore this error." s3sessions/s3handler.go:341
```

Setting `AWS_USE_FIPS_ENDPOINT` appears to work and I did not personally see anything weird with DNS. The following Wireshark logs are a snippet of a script of session creations that I ran and I never saw a lookup for a non FIPS route.
```
sudo tshark -O DNS -T fields -e frame.number -e frame.protocols -e dns.qry.name -e http.host -Y "dns.qry.name or http.host"
Running as user "root" and group "root". This could be dangerous.
Capturing on 'eth0'
 ** (tshark:51566) 02:10:26.354038 [Main MESSAGE] -- Capture started.
 ** (tshark:51566) 02:10:26.354101 [Main MESSAGE] -- File: "/tmp/wireshark_eth0NF7PN1.pcapng"
1477	eth:ethertype:ip:udp:ssdp		239.255.255.250:1900
1479	eth:ethertype:ip:udp:ssdp		239.255.255.250:1900
1548	eth:ethertype:ip:udp:ssdp		239.255.255.250:1900
1724	eth:ethertype:ip:udp:dns	ccooper-fips-bucket.s3-fips.us-east-1.amazonaws.com
1725	eth:ethertype:ip:udp:dns	ccooper-fips-bucket.s3-fips.us-east-1.amazonaws.com
1726	eth:ethertype:ip:udp:dns	ccooper-fips-bucket.s3-fips.us-east-1.amazonaws.com
1727	eth:ethertype:ip:udp:dns	ccooper-fips-bucket.s3-fips.us-east-1.amazonaws.com
1728	eth:ethertype:ip:udp:dns	s3-fips-r-w.us-east-1.amazonaws.com
1729	eth:ethertype:ip:udp:dns	s3-fips-r-w.us-east-1.amazonaws.com
2283	eth:ethertype:ip:udp:ssdp		239.255.255.250:1900
2285	eth:ethertype:ip:udp:ssdp		239.255.255.250:1900
2453	eth:ethertype:ip:udp:ssdp		239.255.255.250:1900
2559	eth:ethertype:ip:udp:dns	ccooper-fips-bucket.s3-fips.us-east-1.amazonaws.com
2560	eth:ethertype:ip:udp:dns	ccooper-fips-bucket.s3-fips.us-east-1.amazonaws.com
2562	eth:ethertype:ip:udp:dns	ccooper-fips-bucket.s3-fips.us-east-1.amazonaws.com
2563	eth:ethertype:ip:udp:dns	ccooper-fips-bucket.s3-fips.us-east-1.amazonaws.com
2564	eth:ethertype:ip:udp:dns	s3-fips-r-w.us-east-1.amazonaws.com
2565	eth:ethertype:ip:udp:dns	s3-fips-r-w.us-east-1.amazonaws.com
3024	eth:ethertype:ip:udp:ssdp		239.255.255.250:1900
3025	eth:ethertype:ip:udp:ssdp		239.255.255.250:1900
3078	eth:ethertype:ip:udp:ssdp		239.255.255.250:1900
3287	eth:ethertype:ip:udp:dns	ccooper-fips-bucket.s3-fips.us-east-1.amazonaws.com
3288	eth:ethertype:ip:udp:dns	ccooper-fips-bucket.s3-fips.us-east-1.amazonaws.com
3290	eth:ethertype:ip:udp:dns	ccooper-fips-bucket.s3-fips.us-east-1.amazonaws.com
3291	eth:ethertype:ip:udp:dns	ccooper-fips-bucket.s3-fips.us-east-1.amazonaws.com
3292	eth:ethertype:ip:udp:dns	s3-fips-r-w.us-east-1.amazonaws.com
3293	eth:ethertype:ip:udp:dns	s3-fips-r-w.us-east-1.amazonaws.com
3610	eth:ethertype:ip:udp:ssdp		239.255.255.250:1900
3611	eth:ethertype:ip:udp:ssdp		239.255.255.250:1900
3646	eth:ethertype:ip:udp:ssdp		239.255.255.250:1900
...
```

Corresponding logs from Teleport all look like this: 
```
...
2022-06-10T02:09:35Z INFO [S3]        Upload created in 306.114342ms. s3sessions/s3stream.go:42
2022-06-10T02:09:35Z INFO [S3]        UploadPart(gJETxQzs96lL1KUfK4prkP401i9wAPC.JztIvmno4lEggvB4NGsZINT_yapxnLiCa7do22EGbTyLrNYWyLJUdJhXYGdd4HvFFWuTp2Di4Y_RXLKEMEh1xYJzgzFmAOHw) part(1) uploaded in 106.158969ms. s3sessions/s3stream.go:70
2022-06-10T02:09:35Z INFO [S3]        UploadPart(gJETxQzs96lL1KUfK4prkP401i9wAPC.JztIvmno4lEggvB4NGsZINT_yapxnLiCa7do22EGbTyLrNYWyLJUdJhXYGdd4HvFFWuTp2Di4Y_RXLKEMEh1xYJzgzFmAOHw) completed in 98.77753ms. s3sessions/s3stream.go:114
2022-06-10T02:09:35Z INFO [AUDIT]     session.upload cluster_name:hephaestus.alistanis.github.beta.tailscale.net code:T2005I ei:2.147483647e+09 event:session.upload sid:0ac06e49-58b1-4357-a7a7-dff698bd3aec time:2022-06-10T02:09:35.329Z url:s3://ccooper-fips-bucket/0ac06e49-58b1-4357-a7a7-dff698bd3aec events/emitter.go:263
...
```

The above url `url:s3://ccooper-fips-bucket/0ac06e49-58b1-4357-a7a7-dff698bd3aec` is translated by the AWS SDK for us.

In summary, my research indicates that the AWS SDK environment variable _should_ be enough for Teleport to do the right thing here. That said, I still think it's a good idea to implement the changes suggested to be explicit and to make the behavior predictable.

I'll follow up soon with PRs addressing those improvements; I'm about 75% of the way there and need to add some tests.

 @pschisa @r0mant  regarding the "strange dns" lookups - is there any information anyone can provide me about the customer's environment and setup?

**Comment by alistanis** [comment]
I removed the endpoint notation, the endpoint notation will make us use the wrong address because it will attach the bucket to the end of the uri.

```
version: v2
teleport:
  storage:
    region: us-east-1
    audit_sessions_uri: "s3://ccooper-fips-bucket/audit_sessions"
```

**Comment by pschisa** [comment]
Quick update, confirmed the issue with the customer environment was that the env variable was not actually being set.

**Comment by pschisa** [comment]
Will this also be backported to v9 and v8 @alistanis ?

**Comment by pschisa** [comment]
confirmed the v9 backport will be fine, no need for v8 currently

Any ETA on getting the v9 backport merged @alistanis ?

**Comment by pschisa** [comment]
@alistanis saw the v9 changes got published, thanks!  Is there any corresponding public documentation I can refer users to in how we ended up implementing this?

**Comment by alistanis** [comment]
@pschisa I have a PR up here for the master documentation: https://github.com/gravitational/teleport/pull/14184 - that'll get backported and adjusted into v9 as well as soon as it's merged into master.
